Add `eth-tester` backend

pcppcp opened the pull request:
- adds `tester` blockchain backend that allows test being run with
  `eth-tester`/`py-evm` combo (use `pytest --blockchain-backend=tester` )
- some test fixtures reorganization
- travis tests still use `geth`

check before merging:
- [x] verify https://github.com/ethereum/eth-tester/pull/100 has been merged & published
- [x] replace `eth-tester` url in `requirements.txt` with PyPi library
- [x] remove git link replacement in `setup.py`

[skip ci]

pcppcp commented on raiden/storage/wal.py:
```diff
@@ -7,6 +7,8 @@
     'InternalEvent',
     ('identifier', 'state_change_id', 'block_number', 'event_object'),
 )
+import structlog
```
not needed really, I just forgot to remove the import after I've removed logs I had in there.

palango replied:
still applies

palango commented on .travis.yml:
```diff
@@ -33,7 +33,7 @@ install:
   ./.travis/before_script.sh
 script:
-  - coverage run -m py.test -Wd --travis-fold=always -vvvvvv --log-config='raiden:DEBUG' --random --blockchain-type=geth $TEST
+  - coverage run -m py.test -Wd --travis-fold=always -xvvvvvv --log-config='raiden:DEBUG' --random --blockchain-type=geth $TEST
```
the `-x` should be removed

palango commented on requirements-dev.txt:
```diff
@@ -24,3 +24,7 @@ releases==1.5.0
 bumpversion==0.5.3
+
+# eth-tester
+git+https://github.com/pcppcp/eth-tester.git@add_status#egg=eth-tester
```
the version with your fixes was released, link to the one on pypi

palango commented on requirements-dev.txt:
```diff
@@ -24,3 +24,7 @@ releases==1.5.0
+
+# eth-tester
+git+https://github.com/pcppcp/eth-tester.git@add_status#egg=eth-tester
+py-evm==0.2.0a18
```
when using eth-tester[py-evm] this shouldn't be required.

pcppcp replied:
evm 0.2.0a26 wasn't compatible with py-evm, but apparently they've fixed it in the new release... I'll update the requirements.

palango commented on raiden/transfer/channel.py:
```diff
@@ -3,6 +3,9 @@
 from collections import namedtuple
+import structlog
+
+log = structlog.get_logger(__name__)  # pylint: disable=invalid-name
```
forgotten as well?

palango commented on raiden/transfer/architecture.py:
```diff
@@ -3,6 +3,10 @@
+import structlog
+
+log = structlog.get_logger(__name__)  # pylint: disable=invalid-name
+
```
forgotten as well?

palango commented on raiden/blockchain/events.py:
```diff
@@ -57,12 +63,16 @@ def get_contract_events(
-        from_block=from_block,
-        to_block=to_block
-    )
+    try:
```
I think we move this into the client, otherwise we have to repeat it everywhere the `get_filter_events`  method is used.

pcppcp replied:
ah, you are right, that makes much more sense.

palango commented on raiden/network/rpc/client.py:
```diff
@@ -186,6 +186,7 @@ def __init__(
             gasprice: int = None,
             nonce_update_interval: float = 5.0,
             nonce_offset: int = 0,
+            web3: Web3 = None
```
you added a lot of commas, but miss one here

pcppcp replied:
I've now added another _large_ bag of commas. All hail the glorious `flake8-commas`.

palango replied:
Hehe, unfortunately there's no auto-fix with that :(

palango commented on raiden/network/rpc/client.py:
```diff
@@ -221,13 +222,20 @@ def __init__(
+            self.web3: Web3 = Web3(HTTPProvider(endpoint))
+        else:
+            self.web3 = web3
+        try:
```
why is this `try` clause necessary

pcppcp replied:
Web3 raises a value error if you are trying to inject middleware that already exists (which is what happens with `tester` - `web3` fixture is session scoped)

palango commented on raiden/network/rpc/client.py:
```diff
@@ -221,13 +222,20 @@ def __init__(
+            pass
+
+        # create the connection test middleware (but only for non-tester chain)
+        if hasattr(web3, 'testing') is False:
```
I'd prefer `if not hasattr(web3, 'testing')`

palango commented on raiden/tests/integration/fixtures/backend_tester.py:
```diff
@@ -0,0 +1,59 @@
+
+
+@pytest.fixture
+def spawn_autominer(web3):
```
Why is this needed? Is the mining blocking otherwise?

pcppcp replied:
The tester only mines a block once you submit a transaction, but no empty blocks are mined. Block generation is required to confirm the transactions.

pcppcp commented:
@palango addressed your comments and rebased to current `master`

LefterisJP commented:
@pcppcp some tests have failed in the latest run

pcppcp commented:
@LefterisJP that's matrix stuff @ulope & @jomuel are looking into it. The CI tests still use `geth` as a backend.

LefterisJP commented on raiden/tests/integration/rpc/test_assumptions.py:
```diff
@@ -93,18 +113,18 @@ def test_estimate_gas_fail(deploy_client, blockchain_backend):
-    host = '0.0.0.0'  # hardcoded in the deploy_client fixture
     second_client = JSONRPCClient(
-        host,
+        '',
```
why did you remove the `'0.0.0.0'` host? Is empty string the same? If yes why remove it and not just be explicit?

pcppcp replied:
`host` and `port` are not used if `web3` is provided. I was thinking of removing those two variables completely and make `web3` a required argument, but I thought I'd address this in a separate PR as it also affects other parts of the code.

LefterisJP replied:
Hmm I am not following.

Web3 is initialized in the ctor of `JSONRPCClient`:

```
        endpoint = 'http://{}:{}'.format(host, port)
        # web3
        self.web3: Web3 = Web3(HTTPProvider(endpoint))
```

pcppcp replied:
you are probably looking at a wrong branch:

https://github.com/pcppcp/raiden/blob/tester_test/raiden/network/rpc/client.py#L231

LefterisJP replied:
Ah okay indeed. Now I see. So you want to make passing the web3 argument mandatory and then get rid of these args. Cool. This can happen in another PR ... but still in the meantime makes no sense to remove `host` from here in this PR.

LefterisJP commented on raiden/tests/unit/conftest.py:
```diff
@@ -0,0 +1,2 @@
+# -*- coding: utf-8 -*-
+# from raiden.tests.integration.fixtures.blockchain import *  # noqa: F401,F403
```
What's this file for? It just seems to have 1 commented import line and nothing else

pcppcp replied:
that's slipped in when rebasing the proxy unit tests. It can go of course.

LefterisJP commented on raiden/tests/utils/blockchain.py:
```diff
@@ -38,6 +38,7 @@ def wait_until_block(chain, block):
     curr_block = chain.block_number()
     while curr_block < block:
         curr_block = chain.next_block()
+        gevent.sleep(0)
```
haha ... I used to have this here but removed it when we removed tester the last time. Are we sure that the `0` value allows for a context switch?

Edit: According to the [docs](http://www.gevent.org/api/gevent.html#gevent.sleep)

> In the current implementation, a value of 0 (the default) means to yield execution to any other runnable greenlets, but this greenlet may be scheduled again before the event loop cycles (in an extreme case, a greenlet that repeatedly sleeps with 0 can prevent greenlets that are ready to do I/O from being scheduled for some (small) period of time); a value greater than 0, on the other hand, will delay running this greenlet until the next iteration of the loop.

So perhaps just switch to a very small positive value instead?

pcppcp replied:
yep, context switch. I was getting infinite loop with tester in there and this seems to help. Good point referring to the docs, I can put something like 1 ms as delay.

LefterisJP commented on raiden/transfer/node.py:
```diff
@@ -294,8 +294,7 @@ def subdispatch_targettask(
         is_valid_subtask = (
-            token_network_identifier == sub_task.token_network_identifier and
-            token_address == sub_task.token_address
+            token_network_identifier == sub_task.token_network_identifier
```
Why did you remove the `token_address` check in the state machine here? What does this code have to do with eth-tester?

pcppcp replied:
that's a bug in Raiden - `token_address` is no longer part of the `TargetTask` structure, accessing it will cause an exception.

LefterisJP replied:
Hmm I see, yes we missed that one in commit: 0bee56ffca77d72540d808b59806ded434ec013d

The initiator and mediator task had it removed. Can you also remove `token_address` from the function and from where it's called up the chain as it's not needed?

LefterisJP commented on setup.cfg:
```diff
@@ -1,10 +1,10 @@
 [flake8]
-ignore = E731, E402, W504
+ignore = E731, E402, W504, C814
```
are you use 814 is the right number. In the docs of the flake8 package it [has](https://github.com/PyCQA/flake8-commas#errors) many similar errors listed and the most generic one sounds like 812

LefterisJP commented on requirements-dev.txt:
```diff
@@ -25,3 +25,7 @@ releases==1.5.0
 # Release
 bumpversion==0.5.3
+
+# eth-tester
```
do you need the commented out lines?

pcppcp replied:
hm, no, I've commented these out when testing dependencies, but since `eth-tester` updated to support most recent `py-evm` the uncommented line is correct and I'll remove the others

pcppcp commented on raiden/tests/integration/fixtures/raiden_network.py:
```diff
@@ -266,6 +277,9 @@ def raiden_network(
     # Force blocknumber update
+    exception = RuntimeError('Alarm failed to start and set up start_block correctly')
+    with gevent.Timeout(seconds=5, exception=exception):
+        wait_for_alarm_start(raiden_apps)
```
the problem is Alarm task.
immediately after the alarm is started, it tries to get the block number it was started in
`self.last_block_number = self.chain.block_number()`
But if context switch due to i/o block occurs at this point, `last_block_number` is still `None`
If you take a look at `raiden_network` fixture, it calls `poll_for_new_block()` immediately after the channels are ready. If the alarm is still waiting for block, the `poll_for_new_block()` method will attempt to perform arithmetic with `last_block_number` still set to `None`

pcppcp commented on raiden/tasks.py:
```diff
@@ -77,6 +77,7 @@ def _run(self):  # pylint: disable=method-hidden
 
     def poll_for_new_block(self):
         current_block = self.chain.block_number()
+        log.info('block', block=current_block)
```
debug leftovers..

pcppcp commented on raiden/tests/integration/contracts/test_secret_registry.py:
```diff
@@ -5,10 +5,10 @@
 def test_secret_registry(secret_registry_proxy):
     #  register secret
     secret = get_random_bytes(32)
+    event_filter = secret_registry_proxy.secret_registered_filter()
```
filter failed to log events if it is created after the event occurs. Not sure why.

pcppcp commented:
@LefterisJP addressed your comments with a new commit

> General Question: After this PR is merged we wil also be able to provide --blockchain-type tester for faster tests locally. But in travis it always remains as geth for the moment. Right?

yes, CI tests are still using `geth` as a backend. But we should consider switching at least non-critical integration tests that require blockchain to `tester`. Contract proxies would be first candidate, followed by various transfer tests.

pcppcp commented on raiden/tests/integration/rpc/test_assumptions.py:
```diff
@@ -9,6 +9,16 @@
 from raiden.network.rpc.client import JSONRPCClient
 from raiden.utils.solc import compile_files_cwd
 
+try:
```
This is related to how `eth-tester` works - it calls the evm directly on executing contract's `call()`, `estimateGas()` or `transfer()` . If `require` is triggered, it'll raise an exception.
Two exceptions that I am trying to import are raised only when using `eth-tester` as a backend. And because tester is a dev dependency, this makes (and fakes) the exceptions to make python happy in case the library isn't available.
